### PR TITLE
✨  Add externally managed annotation and predicate

### PR DESCRIPTION
**Pull request by JoelSpeed**
**What this PR does / why we need it**:

This adds the externally managed annotation that can be added to any InfraCluster resource to mark it as externally managed. InfraCluster controllers should leverage the new predicate to filter events from resources that have been marked as externally managed. This will prevent them from reconciling the resource.

This implements the proposal described in #4135

### CODE REVIEW

File: util/predicates/generic_predicates.go
Comment (alexander-demichev) [inline util/predicates/generic_predicates.go:213]: typo: `predcate`

File: api/v1alpha4/common_types.go
Comment (CecileRobertMichon) [inline api/v1alpha4/common_types.go:87]:
```suggestion
	// An external controller must fulfill the contract of the InfraCluster resource.
```

File: api/v1alpha4/common_types.go
Comment (CecileRobertMichon) [inline api/v1alpha4/common_types.go:86]: should this assumption be documented in the provider contract?

File: util/annotations/helpers.go
Comment (vincepri) [inline util/annotations/helpers.go:35]:
```suggestion
func IsManagedExternally(o metav1.Object) bool {
```

File: api/v1alpha4/common_types.go
Comment (ashish-amarnath) [inline api/v1alpha4/common_types.go:83]:
```suggestion
	// ExternallyManagedAnnotation is an annotation that can be applied to InfraCluster CRs to signify that
```

Comment (ashish-amarnath) [reply api/v1alpha4/common_types.go:83]: This will clarify that this is referring to InfraCluster CRs and not cloud infra resources

File: util/predicates/generic_predicates.go
Comment (furkatgofurov7) [inline util/predicates/generic_predicates.go:237]: Considering if only [above](https://github.com/kubernetes-sigs/cluster-api/pull/4303/files#r593488906) suggestion is taken
```suggestion
	if annotations.IsManagedExternally(obj) {
```

File: api/v1alpha4/common_types.go
Comment (vincepri) [reply api/v1alpha4/common_types.go:83]: Do we need to use an acronym here? `resources` is clear imo

Comment (JoelSpeed) [reply api/v1alpha4/common_types.go:83]: I also think this is clearer as `resources`, gonna leave this one as is

File: api/v1alpha4/common_types.go
Comment (JoelSpeed) [reply api/v1alpha4/common_types.go:86]: Just to confirm, that means adding it to https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/book/src/developer/providers/cluster-infrastructure.md? I've updated the diagram there and added a note to the top of the normal resource flow bit.

Is it worth adding a more detailed section about the externally managed infrastructure in here, or does that belong somewhere else in the docs?

File: util/predicates/generic_predicates.go
Comment (enxebre) [inline util/predicates/generic_predicates.go:235]: Couldn't we just inline `obj.Kind` in 236?

Comment (JoelSpeed) [reply util/predicates/generic_predicates.go:235]: AFAIK `client.Object` is just an interface of `metav1.Object` and `runtime.Object` so we have to use the accessors built into these interfaces, in this case, `runtime.Object`. This is the simplest way to get the `Kind` string out that I'm aware of when using the interfaces

**Comment by JoelSpeed** [comment]
/hold cancel

The proposal is ready to merge (pending lazy consensus) so I think we can get this merged too

File: util/predicates/generic_predicates.go
Comment (enxebre) [inline util/predicates/generic_predicates.go:236]: key/value `kind`/`obj.Name` does not seems correct?
I'd expect either `name/obj.Name` or pass the e.meta from the inner function and use `kind/meta.Kind` here

Comment (JoelSpeed) [reply util/predicates/generic_predicates.go:236]: I copied this line from L#204, I think the intention is that you can use the kind as the key so that it's dynamic for the different controllers

**Comment by enxebre** [comment]
Thanks!
/lgtm

**Review by alexander-demichev — APPROVED** [review]
thanks! The related proposal was merged

File: api/v1alpha4/common_types.go
Comment (CecileRobertMichon) [reply api/v1alpha4/common_types.go:90]: can you also add it to https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/book/src/developer/providers/v1alpha3-to-v1alpha4.md with required changes for providers?

Comment (JoelSpeed) [reply api/v1alpha4/common_types.go:90]: @CecileRobertMichon I've added a small section to the bottom of the document. Let me know what you think, I am happy to add more context if you think it's required

Comment (CecileRobertMichon) [reply api/v1alpha4/common_types.go:90]: looks great, thanks

File: api/v1alpha4/common_types.go
Comment (vincepri) [inline api/v1alpha4/common_types.go:93]: Should this be `ManagedByAnnotation`?

**Review by vincepri — CHANGES_REQUESTED** [review]

**Comment by JoelSpeed** [comment]
@vincepri @sbueringer Thanks for the feedback, I've gone through and made everything consistent now to be `ManagedBy` for the annotation and `ManagedExternally` for the predicates etc. I think I got all the cases

File: docs/book/src/developer/providers/v1alpha3-to-v1alpha4.md
Comment (sbueringer) [reply docs/book/src/developer/providers/v1alpha3-to-v1alpha4.md:261]: Hm the question really is, what are the alternatives. Not sure if `ResourceManagedByUs` or something similar to get rid of the `not` would be better.

**Comment by sbueringer** [comment]
@JoelSpeed let's see if somebody comes up with a better idea: https://github.com/kubernetes-sigs/cluster-api/pull/4303#discussion_r623946620
If not, lgtm from me.

File: docs/book/src/developer/providers/v1alpha3-to-v1alpha4.md
Comment (JoelSpeed) [reply docs/book/src/developer/providers/v1alpha3-to-v1alpha4.md:261]: I was following the `ResourceNotPaused` approach for the naming here. IMO the problem is the same for that one, "with event filter resource not paused". Happy to rename if we can think of something better.

- ResourceIsManaged
- ResourceInternallyManaged

🤔

Comment (sbueringer) [reply docs/book/src/developer/providers/v1alpha3-to-v1alpha4.md:261]: looking at: https://cs.k8s.io/?q=WithEventFilter&i=nope&files=&excludeFiles=&repos= 
Sometimes the predicates are also named in this style: `WithEventFilter(resource.NewPredicates(isInOperatorNamespace)).` Ideally we shorten that even further.

So in our case this could be one of:
* WithEventFilter(resource.IsNotManagedExternally)
* WithEventFilter(resource.IsManagedByUs)
* WithEventFilter(resource.IsManaged)
* WithEventFilter(resource.IsInternallyManaged)

I like about that that it does move the "Resource" in the package path, which removes the "useless" `predicates`. Not sure if I like the implications on the go package names and the deprecations this requires.

Comment (vincepri) [reply docs/book/src/developer/providers/v1alpha3-to-v1alpha4.md:261]: From above `WithEventFilter(resource.IsNotManagedExternally)`, I would tweak it a bit with `WithEventFilter(resource.IsNotExternallyManaged)`? That reads much better, thoughts?

Comment (sbueringer) [reply docs/book/src/developer/providers/v1alpha3-to-v1alpha4.md:261]: I recommended https://github.com/kubernetes-sigs/cluster-api/pull/4303#discussion_r618636981 to turn it around because we already had `IsManagedExternally`. I think we should keep both consistent. English is not my first language, but to me `WithEventFilter(resource.IsNotExternallyManaged)` also sounds more fluent.

Comment (JoelSpeed) [reply docs/book/src/developer/providers/v1alpha3-to-v1alpha4.md:261]: Yeah we had a mix of `ExternallyManaged` and `ManagedExternally` so I made that consistent after @sbueringer highlighted, I don't have a strong opinion on which direction is better though, they both work IMO. Happy to switch the other way but if we do I'll do it across all the places where we have these two words together.

Note the proposal was always "externally managed", so maybe we want to go that way? `Managed Externally Infrastructure` doesn't work, `Externally Managed Infrastructure` does 🤔

Comment (vincepri) [reply docs/book/src/developer/providers/v1alpha3-to-v1alpha4.md:261]: Maybe let's just do `ResourceIsNotExternallyManaged`? Reads better with the `is` in it

**Comment by JoelSpeed** [comment]
@sbueringer @vincepri I've pushed an update to reflect the naming discussion we were having, PTAL

**Review by vincepri — APPROVED** [review]
/approve

**Pull request merged**
